### DISCUSSION TITLE
learn-js/jsintro_pt3

### DISCUSSION THREAD

**Discussion by giscus** [ct-js/docs.ctjs.rocks#135] [Pages' discussions]
# learn-js/jsintro_pt3

Introduction to JavaScript, part III: Objects and Arrays, in-depth Objects Everything in JavaScript is an object! Except for true, false, simple numbers and strings, null, and u...

https://docs.ctjs.rocks/learn-js/jsintro_pt3.html

**Comment by netluke** [comment]
In the second example of the Array's reduce method, I get the following error from the compiler:

Array.prototype.reduce() expects a return value from arrow function.

var monstersInTheLevel = waves.reduce((currentStats, wave) => {
    for (const monsterGroup of wave.monsters) {
        // Initialize a monster group if it hasn't been written to the stats object yet
        if (!currentStats[monsterGroup.type]) {
            currentStats[monsterGroup.type] = 0;
        }
        currentStats[monsterGroup.type] += monsterGroup.amount;
    }
}, {}); // Start with an empty object

The solution is to add a return to the end of the internal anonymous function:

var monstersInTheLevel = waves.reduce((currentStats, wave) => {
    for (const monsterGroup of wave.monsters) {
        // Initialize a monster group if it hasn't been written to the stats object yet
        if (!currentStats[monsterGroup.type]) {
            currentStats[monsterGroup.type] = 0;
        }
        currentStats[monsterGroup.type] += monsterGroup.amount;
    }

   return currentStats;  // THE MISSING PIECE.

}, {}); // Start with an empty object
